hack/README.md

```diff
@@ -0,0 +1,215 @@
+  3. [Implementation Comments](#implementation-comments)
+4. [Formatting](#formatting)
+  1. [Indentation](#indentation)
+  2. [Line Length](#line-length)
+  3. [Pipelines](#pipelines)
+  4. [Loops](#loops)
+  5. [Variable Expansion](#variable-expansion)
+  6. [Quoting](#quoting)
+5. [Other Guidelines](#other-guidelines)
+  1. [Shell Script Safety](#shell-script-safety)
+  2. [Passing Arguments](#passing-arguments)
+  3. [Command Substitution](#command-substitution)
+  4. [Tests](#tests)
+  5. [Testing Strings](#testing-strings)
+  6. [Eval](#eval)
+  7. [Pipes to While](#pipes-to-while)
+6. [Naming Conventions](#naming-conventions)
+  1. [Functions](#functions)
+  2. [Variable Names](#variable-names)
+  3. [Readonly Variables](#readonly-variables)
+  4. [Local Variables](#local-variables)
+
+## Introduction
+
+### Which Shell To Use
+Bash should be the only shell used for OpenShift scripts. 
+
+### When to use Shell
+Shell scripts should only be written when creating small utility functions or wrapping executable functionality.
+
+If a script begins to use arrays or other "advanced" shell features, it should be re-written in Go.
+
+If a script gets longer than several hundred lines, it should be re-written in Go.
+
+## Output
+
+### Standard Out and Standard Error
+Scripts should be careful to write to `stdout` whenever they are writing an error message, and `stdout` when they are not.
+
+## Comments
+
+### Header
+Every file should begin with a short description of its contents. If additions are being made and this requires editing the header comment, it's most likely that the additions should have been made to their own file.
+
+Example:
+```sh
+#!/bin/bash
+
+# This file contains helpful aliases for manipulating the output text to the terminal as
+# well as functions for one-command augmented printing.
+```
+
+### Function Comments
+Any function that is either non-obvious or long should be documented. Any function in a library should be documented.
+
+All function comments must contain:
+ * the full name of the function (including namespaces) as the first entry on the first line of the function
+ * a description of the function
+ * a list of global variables used in the function
+ * a list of arguments to the function
+ * what the function returns, if it returns something other than the exit status of the last command run
+
+Example:
+```sh
+# os::cmd::internal::expect_exit_code_run_grep runs a command and makes an assertion about its exit code
+# and the contents of the command's output to 'stdout' and/or 'stderr' Output from the command is suppressed
+# unless either `VERBOSE=1` or the test fails. 
+# Globals:
+#  - VERBOSE
+# Arguments:
+#  - 1: the command to run
+#  - 2: the command result code evaluation function
+#  - 3: the arguments to pass to 'grep'
+#  - 4: the command output evaluation function
+# Return:
+#  0 if all assertions are met for the command
+#  1 if any assertion is not met for he command
+```
```

Comment:
Maybe we should add yet another example with the function without argument that returns nothing?